### PR TITLE
Use Erlang logger handlers in ExUnit.CaptureLog

### PR DESCRIPTION
**Pull request by whatyouhide**
Part of #9465.

### CODE REVIEW

**Comment by albertored** [comment]
I'm experiencing a strange behavior related to this, logs from erlang libraries in my elixir project are not captured by ExUnit and show up in the console while running tests. Is that known and is this the problem this PR is solving?

(@whatyouhide sorry for the noise but I just found this problem today and this PR "è arrivata a fagiuolo" :smile:  )

File: lib/ex_unit/lib/ex_unit/capture_server.ex [hunk 1]
```diff
@@ -59,14 +66,18 @@ defmodule ExUnit.CaptureServer do
...
+    true = :ets.insert(@ets, {ref, string_io})
 
     if map_size(refs) == 1 do
+      handler_cfg = %{formatter: Logger.default_formatter()}
+
       status =
         with {:ok, config} <- :logger.get_handler_config(:default),
-             :ok <- :logger.remove_handler(:default) do
+             :ok <- :logger.remove_handler(:default),
+             :ok <- :logger.add_handler(@name, __MODULE__, handler_cfg) do
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/capture_server.ex:80]: We should attach this handler even if there is no default. So I recommend moving it out of the `with` and keeping these two code paths separated.

Comment (josevalim) [reply lib/ex_unit/lib/ex_unit/capture_server.ex:80]: We should also remember to remove this handler on maybe_add_console.

Comment (josevalim) [reply lib/ex_unit/lib/ex_unit/capture_server.ex:80]: maybe_add_console should probably be renamed to revert_handlers or something :D

File: lib/ex_unit/lib/ex_unit/capture_log.ex [hunk 2]
```diff
@@ -93,16 +93,17 @@ defmodule ExUnit.CaptureLog do
...
   @doc since: "1.13.0"
-  @spec with_log(keyword, (-> any)) :: {any, String.t()}
-  def with_log(opts \\ [], fun) do
+  @spec with_log(keyword, (-> result)) :: {result, String.t()} when result: any
+  def with_log(opts \\ [], fun) when is_list(opts) do
     opts = Keyword.put_new(opts, :level, nil)
     {:ok, string_io} = StringIO.open("")
 
     try do
       :ok = add_capture(string_io, opts)
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/capture_log.ex:105]: We should remove this capture. We are likely capturing everything twice right now. :) plus, we need to pass the level to the ets table and only write if it matches the ETS table level.

Comment (josevalim) [reply lib/ex_unit/lib/ex_unit/capture_log.ex:105]: Note we can default the level to :all.

**Comment by whatyouhide** [comment]
@josevalim we had the same ideas, yep. I just implemented all of that, ready for another review!

File: lib/ex_unit/lib/ex_unit/capture_server.ex [hunk 3]
```diff
@@ -59,11 +66,17 @@ defmodule ExUnit.CaptureServer do
     {:reply, :ok, release_device(ref, config)}
   end
 
-  def handle_call({:log_capture_on, pid}, _from, config) do
+  def handle_call({:log_capture_on, pid, string_io, opts}, _from, config) do
     ref = Process.monitor(pid)
     refs = Map.put(config.log_captures, ref, true)
```
Comment (whatyouhide) [inline lib/ex_unit/lib/ex_unit/capture_server.ex:71]: I think we'll be able to get rid of this and only use the ETS table to store these, but I'd rather do one thing at a time and do that in another PR. 🙃

File: lib/ex_unit/lib/ex_unit/capture_server.ex [hunk 4]
```diff
@@ -186,21 +200,37 @@ defmodule ExUnit.CaptureServer do
...
         config
     end
   end
 
-  defp maybe_add_console(refs, {:ok, %{module: module} = config}) when map_size(refs) == 0 do
+  defp maybe_revert_to_default_handler(refs, {:ok, %{module: module} = config})
+       when map_size(refs) == 0 do
+    :logger.remove_handler(@name)
     :logger.add_handler(:default, module, config)
   end
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/capture_server.ex:219]:
```suggestion
  defp maybe_revert_to_default_handler(refs, status) when map_size(refs) == 0 do
    :logger.remove_handler(@name)

    with {:ok, %{module: module} = config} <- status do
      :logger.add_handler(:default, module, config)
    end
  end
```

**Review by josevalim — APPROVED** [review]
One tiny comment and ship it!

**Review by josevalim — APPROVED** [review]

File: lib/ex_unit/test/ex_unit/capture_log_test.exs [hunk 5]
```diff
@@ -78,24 +78,39 @@ defmodule ExUnit.CaptureLogTest do
...
+    end
   end
 
   defp wait_capture_removal() do
-    case :gen_event.which_handlers(Logger) do
-      [Logger.Backends.Config] ->
-        :ok
-
-      _otherwise ->
+    case Enum.map(:logger.get_config().handlers, & &1.id) do
```
Comment (josevalim) [inline lib/ex_unit/test/ex_unit/capture_log_test.exs:107]:
```suggestion
    case Enum.map(:logger.get_handler_config(), & &1.id) do
```
